### PR TITLE
[RAPTOR-3841] Implement predictUnstructured endpoint for DRUM uwsgi

### PR DESCRIPTION
**Pull request by yakov-g**
# This repository is public. Do not put here any private DataRobot or customer's data: code, datasets, model artifacts, .etc.

## Summary
Implement predictUnstructured endpoint for DRUM uwsgi

Also:
- added scipy as dep fro drum (required after sparse addition)
- added `-U` option to `pip install` to pull the latest mlpiper (ask me for more details if you're interested)

### CODE REVIEW

File: custom_model_runner/datarobot_drum/resource/unstructured_helpers.py [hunk 1]
```diff
@@ -1,13 +1,15 @@
...
 CHARSET_DEFAULT = "utf8"
-MIMETYPE_TEXT_DEFAULT = "text/plain"
-MIMETYPE_BINARY_DEFAULT = "application/octet-stream"
+MIMETYPE_TEXT_DEFAULT = PredictionServerMimetypes.TEXT_PLAIN
+MIMETYPE_BINARY_DEFAULT = PredictionServerMimetypes.APPLICATION_OCTET_STREAM
 
 
 def _is_text_mimetype(mimetype):
-    return mimetype.startswith("text/") or mimetype.startswith("application/json")
+    return mimetype.startswith("text/") or mimetype.startswith(
```
Comment (klichukb) [inline custom_model_runner/datarobot_drum/resource/unstructured_helpers.py:10]: 'startswith' is not correct for application/json, it should strictly equal

File: custom_model_runner/datarobot_drum/resource/components/Python/uwsgi_component/uwsgi_serving.py [hunk 2]
```diff
@@ -149,6 +149,28 @@ def predict(self, url_params, form_params):
...
+        methods=["POST"],
+    )
+    def predict_unstructured(self, url_params, form_params):
+        if self._error_response:
+            return HTTP_513_DRUM_PIPELINE_ERROR, self._error_response
+
+        self._stats_collector.enable()
+        self._stats_collector.mark("start")
+
+        response, response_status = self.do_predict_unstructured()
```
Comment (klichukb) [inline custom_model_runner/datarobot_drum/resource/components/Python/uwsgi_component/uwsgi_serving.py:163]: should be wrapped with try finally and mark(finish)+disable in finally block.

File: custom_model_runner/datarobot_drum/resource/unstructured_helpers.py (hunk 1 above)
Comment (yakov-g) [reply custom_model_runner/datarobot_drum/resource/unstructured_helpers.py:10]: ok

File: custom_model_runner/datarobot_drum/resource/components/Python/uwsgi_component/uwsgi_serving.py (hunk 2 above)
Comment (yakov-g) [reply custom_model_runner/datarobot_drum/resource/components/Python/uwsgi_component/uwsgi_serving.py:163]: ok

**Review by klichukb — COMMENTED** [review]
Two code-related comments

**Review by klichukb — APPROVED** [review]

File: model_templates/inference/python3_unstructured/custom.py [hunk 3]
```diff
@@ -12,6 +12,7 @@ def score_unstructured(model, data, query, **kwargs):
     if isinstance(data, bytes):
         data = data.decode("utf8")
 
+    data = data.strip().replace("  ", " ")
     words_count = data.count(" ") + 1
```
Comment (meltzg) [inline model_templates/inference/python3_unstructured/custom.py:16]: nit. could just be `len(data.split())` Default behavior splits on all whitespace

**Review by meltzg — APPROVED** [review]

**Comment by datarobot-jarvis** [comment]
:dollar: Total cost of executing tests in Jenkins: $**4.87** USD.

[What is this all about?](http://bit.ly/2BahK4V)

**Pull request merged**